subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -571,7 +571,7 @@ static int dtls_rx(void *ctx, unsigned char *buf, size_t len,
 		/* mbedtLS does not allow blocking rx for DTLS, therefore use
 		 * k_poll for timeout functionality.
 		 */
-		if (is_block) {
+		if (!is_block) {
```

Comment:
This piece of code is correct, let me try to explain:

mbedTLS has two types of TLS recv handlers: `mbedtls_ssl_recv_t` and `mbedtls_ssl_recv_timeout_t`. For TLS we use the first one, where we use use either blocking or non-blocking `zsock_recvfrom()` call, depending on whether the TLS socket should be blocking or non-blocking. This callback can block indefinitely, just as socket does.

But for DTLS, it's a bit more complex as in the mbedTLS documentation we have:
```
For DTLS, you need to provide either a non-NULL f_recv_timeout callback, or a f_recv that doesn't block.
```
It's likely due to handshake retransmissions that mbedTLS needs to handle on its own. As we still want to have blocking-like behaviour, I've implemented `f_recv_timeout` kind of timeout but it was needed to add some kind of timeout mechanism to a **blocking** socket, which is done with the internal `poll()` call. So it's correct that we use `poll()` only for blocking socket - otherewise we just check for data with non-blocking `zsock_recvfrom()` and return promptly.

This logic could likely be simplified now, as we have `SO_RCVTIMEO` implemented (it was not present at the time the TLS sockets were added), but that's a story for a different PR.

Reply:
@rlubos  Thank you for explaining! Then I misunderstood this piece of code and I'll close this pull request.

I tried to use Ublox Sara-R4 modem with native DTLS stack, however obviously it blocks when it's supposed to be non-blocking which is a bug in itself.